keps/sig-cluster-lifecycle/kubeadm/20191229-kubeadm-move-out-of-kk.md

```diff
@@ -0,0 +1,635 @@
+
+kubeadm no longer imports k/k packages, so this is already possible
+if one just copies cmd/kubeadm outside of k/k and creates a new Go
+module for it. It still requires some hacks around applying a version in the
+kubeadm binary. Once kubeadm moves out both the build tooling and module
+will be available to users for easy building.
+
+#### Story 2
+
+As a user, I would like to import the kubeadm public, versioned API without
+importing kubernetes/kubernetes.
+
+Currently importing the k/k root module will make your build fail.
+The best solution for users is to fork the kubeadm API packages.
+Once kubeadm moves out, importing the "kubeadm" module will make
+it possible to easily import only the API portion.
+
+#### Story 3
+
+As a user, I would like to import utility packages from kubeadm.
+
+Similar to the above user story, importing kubeadm from k/k will fail.
+Once kubeadm moves out, utility functions will be exposed for public use,
+but perhaps without any guarantees until the kubeadm backend is truly
+announced for public consumption.
+
+### Implementation Details/Notes/Constraints
+
+#### Location
+
+The proposed new location is the kubernetes/kubeadm repository.
+
+This makes it possible to use `go get k8s.io/kubeadm` to obtain
+the kubeadm source tree.
+
+The current contents of the `k/k/cmd/kubeadm/app` directory
+should be moved in the root of k/kubeadm repository.
+
+The current `k/k/cmd/kubeadm/test` directory that contains
+integration tests should be merged with the `k/kubeadm/tests`
+directory under `k/kubeadm/tests/integration`.
+
+Existing tools in the k/kubeadm repository such as "kinder" and
+"operator" must continue existing as separate modules (sub-folders)
+inside the kubeadm module and can be obtained using
+`go get k8s.io/kubeadm/some-tool`.
+
+Existing docs and tests must co-exist in sub-folders too.
+
+Side effects:
+- All contents of k/kubeadm will be branched and tagged with kubeadm,
+but this is actually a positive change.
+- `hack/verify*` scripts for the root kubeadm folder should skip "kinder"
+and "operator". This area can see improvements in the future,
+potentially sharing scripts for all modules.
+
+#### Handling older kubeadm versions
+
+Currently kubeadm in the `release-1.17` branch (and later) of k/k
+does not depend on the rest of k/k. This was required for kubeadm
+to move out, however the older branches of k/k include versions of
+kubeadm that still depend on k/k.
+
+The proposed way forward is to move only the latest kubeadm from k/k master
+and build older version of kubeadm from k/k.
+
+This creates the following complications:
+- The post-submit job must handle tags and branches for kubeadm versions
+larger than X.
+- The kubernetes/release tooling must build kubeadm version X from the
+kubeadm repository while build older kubeadm from k/k.
+
+The cherry-pick script from k/k/hack will not work out of the box
+for the kubeadm repository so potentially it can be forked to suite
```

Comment:
I don't like the idea of forking versus making a common tool.  We're going to need this cherry picker for k/release also. 
And then the tool is also branched in each of those repos as we go forward.  If we put it in k/release there'd be just one, but making k/k or k/kubeadm folks also check out k/release and use the right branch's hack/cherry_pick_pull.sh isn't a simplification either.  I'm not sure how we should approach the idea of common tools and lifecycle management of them.  In the short term I guess that may just mean duplicating content.

Reply:
probably should mentioned that my comment there was just an assumption.
i just did some testing and i think the script will just work for k/kubeadm or any other repository because it's based on GIT remotes.

will update this part to note that any changes will go in k/k/hack/cherry_pick_pull.sh making the script re-usable instead of forking.